python/uv/private/uv.bzl

```diff
@@ -0,0 +1,198 @@
+    attrs = {
+        "compatible_with": attr.label_list(
+            doc = "The compatible with constraint values for toolchain resolution",
+        ),
+        "name": attr.string(
+            doc = "The platform string used in the UV repository to denote the platform triple.",
+            mandatory = True,
+        ),
+    },
+)
+
+uv_toolchain = tag_class(
+    doc = "Configure uv toolchain for lock file generation.",
+    attrs = {
+        "name": attr.string(
+            doc = "The name of the toolchain repo",
+            default = "uv_toolchains",
+        ),
+        "version": attr.string(
+            doc = "Explicit version of uv.",
+            mandatory = True,
+        ),
+    },
+)
+
+def _uv_toolchain_extension(module_ctx):
+    config = {
+        "platforms": {},
+    }
+
+    for mod in module_ctx.modules:
+        if not mod.is_root and not mod.name == "rules_python":
+            # Only rules_python and the root module can configure this.
+            #
+            # Ignore any attempts to configure the `uv` toolchain elsewhere
+            #
+            # Only the root module may configure the uv toolchain.
+            # This prevents conflicting registrations with any other modules.
+            #
+            # NOTE: We may wish to enforce a policy where toolchain configuration is only allowed in the root module, or in rules_python. See https://github.com/bazelbuild/bazel/discussions/22024
+            continue
+
+        # Note, that the first registration will always win, giving priority to
+        # the root module.
+
+        for platform_attr in mod.tags.platform:
+            config["platforms"].setdefault(platform_attr.name, struct(
+                name = platform_attr.name.replace("-", "_").lower(),
+                compatible_with = platform_attr.compatible_with,
+            ))
+
+        for config_attr in mod.tags.config:
+            config.setdefault("base_url", config_attr.base_url)
+
+        for toolchain in mod.tags.toolchain:
+            config.setdefault("version", toolchain.version)
+            config.setdefault("name", toolchain.name)
+
+    if not config["version"]:
+        return
+
+    config.setdefault("base_url", _DEFAULT_BASE_URL)
+    config["urls"] = _get_tool_urls_from_dist_manifest(
+        module_ctx,
+        base_url = "{base_url}/{version}".format(**config),
+    )
+    uv_repositories(
+        name = config["name"],
+        platforms = config["platforms"],
+        urls = config["urls"],
+        version = config["version"],
+    )
+
+def _get_tool_urls_from_dist_manifest(module_ctx, *, base_url):
+    """Download the results about remote tool sources.
+
+    This relies on the tools using the cargo packaging to infer the actual
+    sha256 values for each binary.
+    """
+    dist_manifest = module_ctx.path(_DIST_MANIFEST_JSON)
+    module_ctx.download(base_url + "/" + _DIST_MANIFEST_JSON, output = dist_manifest)
```

Comment:
We probably want a sha for the dist manifest download?

Using untrusted shas to verify a download is equivalent to not using shas

Reply:
Regarding the `sha256`, I am downloading the manifest that only has links and I don't see how the extra `sha256` value here could be beneficial.

This has been inspired by https://github.com/bazel-contrib/rules_go/blob/7f6a9bf5870f2b5ffbba1615658676dcabf9edc7/go/private/sdk.bzl#L84 where `rules_go` is downloading things from the manifest if the user does not specify anything.

Maybe we should instead an optional `sha256s` dict where we have platform labels as keys and the sha values as values? And we could print a warning with `buildozer` command to add the said `sha256s` values to their `MODULE.bazel` file if they want things to be deterministic?

Reply:
Ah, I think I see your point now. Hm.

On the one hand, not using shas means, if someone MITM the manifest, then they can control what actual uv stuff is downloaded and then used.

On the other hand, using a sha means we are tied to a particular manifest and don't benefit from the automatic-ness of using the manifest. Which is pretty appealing behavior.

Hm, not sure how, or if, we can split the difference. These seem at odds.

Also:
* Does the sha arg of download() allow bazel to better cache it?
* Does the sha arg feed into MODULE.lock? I would expect so, since this is module-phase behavior.
* Without the sha arg set, does Bazel print a warning? (http_archive does, not sure if download does)